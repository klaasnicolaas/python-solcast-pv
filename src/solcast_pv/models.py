"""Asynchronous Python client for Solcast."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any


@dataclass
class RooftopSite:
    """Object representing the rooftop site for the Solcast API."""

    name: str
    resource_id: str
    install_date: datetime

    capacity: float
    capacity_dc: float
    azimuth: int
    tilt: int
    loss_factor: float

    @classmethod
    def from_dict(cls: type[RooftopSite], data: dict[str, Any]) -> RooftopSite:
        """Create a new instance of the RooftopSites class from a dictionary."""
        return cls(
            name=data["name"],
            resource_id=data["resource_id"],
            install_date=datetime.fromisoformat(data["install_date"]),
            capacity=data["capacity"],
            capacity_dc=data["capacity_dc"],
            azimuth=data["azimuth"],
            tilt=data["tilt"],
            loss_factor=data["loss_factor"],
        )


@dataclass
class RateLimit:
    """Object representing the rate limit status for the Solcast API."""

    daily_limit: int
    remaining_daily: int
    consumed_daily: int

    @classmethod
    def from_dict(cls: type[RateLimit], data: dict[str, Any]) -> RateLimit:
        """Create a new instance of the RateLimit class from a dictionary."""
        return cls(
            daily_limit=data["daily_limit"],
            remaining_daily=int(data["daily_limit"] - data["daily_limit_consumed"]),
            consumed_daily=data["daily_limit_consumed"],
        )
