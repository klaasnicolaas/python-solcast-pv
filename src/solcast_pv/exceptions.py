"""Asynchronous Python client for Solcast."""


class SolcastError(Exception):
    """Generic Solcast exception."""


class SolcastConnectionError(SolcastError):
    """Solcast connection exception."""


class SolcastAuthenticationError(SolcastError):
    """Solcast authentication exception."""


class SolcastResultsError(SolcastError):
    """Solcast results exception."""
