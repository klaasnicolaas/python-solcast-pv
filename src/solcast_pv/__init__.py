"""Asynchronous Python client for Solcast."""

from .exceptions import (
    SolcastAuthenticationError,
    SolcastConnectionError,
    SolcastError,
    SolcastResultsError,
)
from .models import RateLimit, RooftopSite
from .solcast_pv import Solcast

__all__ = [
    "RateLimit",
    "RooftopSite",
    "Solcast",
    "SolcastAuthenticationError",
    "SolcastConnectionError",
    "SolcastError",
    "SolcastResultsError",
]
