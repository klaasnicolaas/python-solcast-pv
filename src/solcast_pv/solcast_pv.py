"""Asynchronous Python client for Solcast."""

from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self

from aiohttp import ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    SolcastAuthenticationError,
    SolcastConnectionError,
    SolcastError,
    SolcastResultsError,
)
from .models import RateLimit, RooftopSite

VERSION = metadata.version(__package__)


@dataclass
class Solcast:
    """Main class for handling connections with the Solcast API."""

    token: str

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Solcast API.

        Args:
        ----
            uri: Request URI, without '/api/', for example, 'status'.
            method: HTTP method to use.
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (JSON decoded) with the response from
            the Solcast API.

        Raises:
        ------
            SolcastConnectionError: Error occurred while connecting to Solcast API.
            SolcastError: Unexpected content type response from Solcast API.

        """
        url = URL.build(
            scheme="https",
            host="api.solcast.com.au",
            path="/",
        ).join(URL(uri))

        headers = {
            "Authorization": f"Bearer {self.token}",
            "Accept": "application/json",
            "User-Agent": f"PythonSolcastPV/{VERSION}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    headers=headers,
                    params=params,
                    ssl=True,
                )
                response.raise_for_status()
        except TimeoutError as exception:
            msg = "Timeout occurred while connecting to Solcast API."
            raise SolcastConnectionError(msg) from exception
        except ClientResponseError as exception:
            if exception.status == 401:
                msg = "Invalid API key provided to Solcast API."
                raise SolcastAuthenticationError(msg) from exception
            if exception.status == 403:
                msg = "API key does not have access to the requested resource."
                raise SolcastAuthenticationError(msg) from exception
            if exception.status == 404:
                msg = "Requested resource was not found on Solcast API."
                raise SolcastError(msg) from exception
            msg = "Error occurred while connecting to Solcast API."
            raise SolcastConnectionError(msg) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while connecting to Solcast API."
            raise SolcastConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from Solcast API."
            raise SolcastError(
                msg,
                {"content_type": content_type, "text": text},
            )

        return await response.json()

    async def get_rooftop_sites(self) -> list[RooftopSite]:
        """Get the rooftop sites for the Solcast API.

        Returns
        -------
            RooftopSite: The rooftop site.

        """
        response = await self._request("rooftop_sites")
        try:
            results: list[RooftopSite] = [
                RooftopSite.from_dict(site) for site in response["sites"]
            ]
        except KeyError as exception:
            msg = "No rooftop sites found on your Solcast account."
            raise SolcastResultsError(msg) from exception
        return results

    async def get_ratelimit(self) -> RateLimit:
        """Get the rate limit status for the Solcast API.

        Returns
        -------
            RateLimit: The rate limit status.

        """
        response = await self._request("json/reply/GetUserUsageAllowance")
        return RateLimit.from_dict(response)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Solcast object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
