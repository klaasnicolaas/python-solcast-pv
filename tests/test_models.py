"""Test the models for Solcast."""

from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from solcast_pv import RateLimit, Solcast

from . import load_fixtures


async def test_rate_limiting(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    solcast_client: Solcast,
) -> None:
    """Test rate limiting."""
    aresponses.add(
        "api.solcast.com.au",
        "/json/reply/GetUserUsageAllowance",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("ratelimit.json"),
        ),
    )
    response: RateLimit = await solcast_client.get_ratelimit()
    assert response == snapshot


async def test_rooftop_sites(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    solcast_client: Solcast,
) -> None:
    """Test rooftop sites."""
    aresponses.add(
        "api.solcast.com.au",
        "/rooftop_sites",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("rooftop_sites.json"),
        ),
    )
    response = await solcast_client.get_rooftop_sites()
    assert response == snapshot
