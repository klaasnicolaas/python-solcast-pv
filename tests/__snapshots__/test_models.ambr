# serializer version: 1
# name: test_rate_limiting
  RateLimit(daily_limit=10, remaining_daily=7, consumed_daily=3)
# ---
# name: test_rooftop_sites
  list([
    RooftopSite(name='Flat roof', resource_id='acqo-rh9b-hq68-3y5n', install_date=datetime.datetime(2023, 5, 11, 22, 0, tzinfo=datetime.timezone.utc), capacity=3.24, capacity_dc=2.64, azimuth=-120, tilt=25, loss_factor=0.98),
    RooftopSite(name='Slanted roof', resource_id='ww00-3og6-dkor-4fy1', install_date=datetime.datetime(2024, 5, 11, 22, 0, tzinfo=datetime.timezone.utc), capacity=1.65, capacity_dc=1.95, azimuth=130, tilt=25, loss_factor=0.98),
  ])
# ---
